src/utils.cpp

```diff
@@ -338,9 +342,44 @@ void Utils::SwapByteOrder(double& d) {
 	p[1] = tmp;
 }
 
+/** Generate a random number in the range [0,max] */
+static uint32_t GetRandomUnsigned(uint32_t max)
+{
+	if (max == 0xffffffffull) return GetRandomU32();
+
+	// Rejection sampling:
+	// 1. Divide the range of uint32 into blocks of max+1
+	//    numbers each, with rem numbers left over.
+	// 2. Generate a random u32. If it belongs to a block,
+	//    mod it into the range [0,max] and accept it.
+	// 3. If it fell into the range of rem leftover numbers,
+	//    reject it and go back to step 2.
+	uint32_t m = max + 1;
+	uint32_t rem = -m % m; // = 2^32 mod m
```

Comment:
Ah now I get it. Didn't knew that it's promoted to unsigned beforehand, interesting trick :D